### ISSUE TITLE
Questions about training code

### ISSUE THREAD

**Issue by LuFan31** [RunpeiDong/DreamLLM#22]
Hi Runpei, 

Thank you for your excellent work.

I use `torchrun -m projects.dreamllm.train --config_file projects/dreamllm/configs/stage1/base.py "training.per_device_train_batch_size=16" "training.output_dir='./work_dirs/your_dir/'" "training.run_name='wandb_run_name'" "training.learning_rate=2e-3"` to train the condition projector. However, when the code reaches https://github.com/RunpeiDong/DreamLLM/blob/master/omni/models/dreamllm/modeling_plugins.py#L243, we found that the loaded file is clip_vision_embedding.bin from the pre-trained DreamLLM weights. But `pretrained_model_name_or_path` is the local path for storing the vicuna7B-v1.1 weights (because the cluster I am using does not have internet access). I want to know which module in the paper clip_vision_embedding.bin refers to, and whether it should be placed in the local path for storing vicuna7B-v1.1 weights.

Thank you for your assistance!

Best
Fan

**Comment by LuFan31** [comment]
And should I set the `local_files_only` mentioned in https://github.com/RunpeiDong/DreamLLM/blob/master/projects/dreamllm/configs/common.py#L7C5-L7C21 as True considering the cluster I am using does not have internet access?

**Comment by RunpeiDong** [comment]
Yes, I think so. If your cluster does not have internet access, then you must set `local_files_only` to True and upload the checkpoint you downloaded from your device to the corresponding folder at the cluster.

**Comment by RunpeiDong** [comment]
Hi @LuFan31,

I notice that you changed your first comment. The `clip_vision_embedding.bin` is actually the CLIP encoder. This is the OpenAI model that you have to download from huggingface. But during training, this .bin file will be also saved.

**Comment by LuFan31** [comment]
But what is the role of the `clip_vision_model_name_or_path` in https://github.com/RunpeiDong/DreamLLM/blob/master/projects/dreamllm/configs/common.py#L30? `clip_vision_embedding.bin` is the weight of CLIP visual encoder and can be download from https://huggingface.co/RunpeiDong/dreamllm-7b-chat-v1.0/tree/main, right? But the `clip_vision_model_name_or_path` is the link of https://huggingface.co/openai/clip-vit-large-patch14/tree/main, where the weight is the whole CLIP, containing both textual and visual encoder. So, does this means that the `clip_vision_model_name_or_path`  does not work during the training process? Because during training, the weights of the CLIP visual encoder are loaded from `clip_vision_embedding.bin` in DreamLLM's huggingface, and there are no separate CLIP visual encoder weights in the OpanAI CLIP's huggingface pointed to by `clip_vision_model_name_or_path`.

**Comment by RunpeiDong** [comment]
You can set this path in the config file. For example, here: https://github.com/RunpeiDong/DreamLLM/blob/fb5f3e06581640cd2123cd50e13fd83f2f7de079/projects/dreamllm/configs/stage2/base.py#L44

If you have loaded the weights from DreamLLM's huggingface repo, you can skip this. It depends on whether you want to unfreeze the ViT model during training. The weight on DreamLLM's huggingface repo is actually the same as OpenAI's model because the model is frozen during training. 

For example, if you unfreeze the CLIP model during stage I training, then in the following stage, you have to use the local path that saves the fine-tuned CLIP model, right? Like the config I show in [DreamLLM/projects/dreamllm/configs/stage2/base.py](https://github.com/RunpeiDong/DreamLLM/blob/fb5f3e06581640cd2123cd50e13fd83f2f7de079/projects/dreamllm/configs/stage2/base.py#L44).

**Comment by LuFan31** [comment]
I plan to reproduce the initial DreamLLM training precess, and freeze the CLIP visual encoder and SD decoder just following you. But some weights need to load make me confused.
**1.** According to your comments, I think the three paths mentioned [here](https://github.com/RunpeiDong/DreamLLM/blob/fb5f3e06581640cd2123cd50e13fd83f2f7de079/projects/dreamllm/configs/stage2/base.py#L43C1-L47C79)
are the path of visual projector, condition projector and learnable dream queries obtained from Stage1 training. And the configs of Stage1 training do not contain these three paths. 
**2.** Factually, we load the weight of CLIP visual encoder, SD decoder using the function `load_model` in Class [CLIPVisionEmbedding](https://github.com/RunpeiDong/DreamLLM/blob/fb5f3e06581640cd2123cd50e13fd83f2f7de079/omni/models/dreamllm/modeling_plugins.py#L297C9-L297C19) and [StableDiffusionHead](https://github.com/RunpeiDong/DreamLLM/blob/fb5f3e06581640cd2123cd50e13fd83f2f7de079/omni/models/dreamllm/modeling_plugins.py#L451C9-L451C19) , right? 
**3.** The `dream_embedding.bin` in DreamLLM's huggingface is the final dream queries trained from the 3 stages mentioned in paper? If so,  I should not load `.bin` for  dream queries when I train the condition projector first in Stage1 (_i.e,_ use the `projects/dreamllm/configs/stage1/base.py` as config), right? The dream queries just need to be initialized and saved after training the condition projector in Stage1, and the saved weight of dream queries need to be loaded during Stage2 training, right?

**Comment by RunpeiDong** [comment]
Hi @LuFan31, here I answer the question one by one:
- A1: Yes.
- A2: Not really. The vision encoder and decoder are first initialized by the default `from_pretrained` method. Then, it will load the weights using `load_model` from the path specified in the config file if it is provided or it can be found in the checkpoint folder. If not, no weights are further loaded, and thus the `load_model` will not work.
- Yes, the `dream_embedding.bin` on the huggingface is the final dream query trained from 3 stages. Yes, then you should not load the weights if you want to train stage 1 from scratch as the LLM is different (stage 1 LLM is the pretrained vicuna while the LLM after stage 2&3 is unfrozen and fine-tuned). Yes, the dream queries should be initialized randomly if you want to train stage 1 from scratch.

**Comment by LuFan31** [comment]
Thank you for your answer! I have understood the issue of loading CLIP visual encoder and SD decoder weights. Next, I will use my own training data format to train DreamLLM.

**Comment by RunpeiDong** [comment]
Geat. Always happy to help!

**Comment by LuFan31** [comment]
Hi Runpei,
 I use `sh scripts/eval/text2img/eval_coco_zero_shot_clip8_select.sh $OUTPUT_DIR $MODEL_NAME_OR_PATH` as mentioned [here](https://github.com/RunpeiDong/DreamLLM/blob/master/omni/eval/README.md?plain=1#L100) for FID evaluation. But the generated icmages are just color blocks:
![image](https://github.com/RunpeiDong/DreamLLM/assets/88334335/3092a1c8-35d5-4e16-89c0-c046b280d7e1)
I have checked and the input prompt is ok.

**Comment by RunpeiDong** [comment]
Hi @LuFan31,

Have you checked if the model has been successfully loaded?

**Comment by LuFan31** [comment]
Yes. I load the weight from [DreamLLM huggingface](https://huggingface.co/RunpeiDong/dreamllm-7b-chat-aesthetic-v1.0/tree/main), and the log information shows that all keys match successfully.

**Comment by RunpeiDong** [comment]
Have you tried the [inference.py](https://github.com/RunpeiDong/DreamLLM/blob/master/projects/dreamllm/inference.py) script? There is one thing that has to be noticed, as shown in the script. The model has to be downloaded to some local folder like:
https://github.com/RunpeiDong/DreamLLM/blob/fb5f3e06581640cd2123cd50e13fd83f2f7de079/projects/dreamllm/inference.py#L36

If you directly load it from huggingface, the model will not be okay. I think it is a bug of hf, but I haven't found a good way to deal with it. So, I think we can just download it using that method for now.

**Comment by LuFan31** [comment]
I have download the DreamLLM weight to a local folder, and set this local path as `model_name_or_path`. And I have run the inference.py, still using the local DreamLLM weight, but the generated images are also color blocks.

**Comment by RunpeiDong** [comment]
You run the inference.py code to download the weight and still got bad results?

**Comment by LuFan31** [comment]
No, I have download the DreamLLM weight to a local folder before running inference.py, and the `model_name_or_path`  [here](https://github.com/RunpeiDong/DreamLLM/blob/master/projects/dreamllm/inference.py#L35) is the local folder, so I skip the `snapshot_download` [here](https://github.com/RunpeiDong/DreamLLM/blob/master/projects/dreamllm/inference.py#L36).

**Comment by RunpeiDong** [comment]
Try to download it using the inference.py script. This is the bug I mentioned. If you directly download it from hf, the model may not work. I don't know why but if you follow the script, it will be okay.

**Comment by LuFan31** [comment]
Sorry, I misunderstand your guidance because I thought running `snapshot_download` is 'directly download it from hf'. I will try to use inference.py script in local devices, considering the cluster I used does not have internet access.

**Comment by FolSpark** [comment]
@RunpeiDong 
When I run inference.py and load `RunpeiDong/dreamllm-7b-chat-v1.0`. Although the inference.py run successfully and output images, but I see INFO below:

![image](https://github.com/user-attachments/assets/9eab7007-1eb0-4033-8867-b4bc5870a69e)

I don't know what this means, is there some weight not be trained?

**Comment by RunpeiDong** [comment]
Hi @ybbwcwaps,

The missing weights seem to be because the implementation of the inv_freq is different. I don't quite remember if this is okay. But I think you can now ignore it since the results are fine. 

For the warning of requiring more training, you can also just ignore it. This is a reminder but I don't think now you have to train mode things.
